### PR TITLE
Detect encoding changes while parsing diff

### LINKED ISSUE #16329: Confusing diff when changing file encoding
- Gitea version (or commit ref): 1.15
- Can you reproduce the bug at https://try.gitea.io:
  - Yes (https://try.gitea.io/jpraet/encoding/commit/7d1a5559f40bb2520c42d08c9508c2e5d82d98d5)

## Description

I have commits that change the encoding of files from Windows Cp1252 encoding to UTF-8.

In Gitea, the diff makes it look like the special characters get converted incorrectly.

https://try.gitea.io/jpraet/encoding/commit/7d1a5559f40bb2520c42d08c9508c2e5d82d98d5
afbeelding

When viewing the new file, it is displayed correctly as UTF-8 though.

https://try.gitea.io/jpraet/encoding/src/commit/7d1a5559f40bb2520c42d08c9508c2e5d82d98d5/test.txt
afbeelding

On GitHub, it looks like this: https://github.com/jpraet/encoding/commit/7d1a5559f40bb2520c42d08c9508c2e5d82d98d5

afbeelding

Git diff output:
afbeelding

### PR DESCRIPTION
**Pull request by jpraet**
Fix #16329

Independently detect encoding of added, deleted and unchanged sections of each file diff.

Before:

afbeelding

After:

afbeelding

### CODE REVIEW

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
You're going to need to add a marker to the file header to say the charsets for the files - otherwise you're likely to get an apparently empty diff on re-encoding.

**Comment by jpraet** [comment]
> 
> 
> You're going to need to add a marker to the file header to say the charsets for the files - otherwise you're likely to get an apparently empty diff on re-encoding.

Indeed you can now have something like this:

image

I agree that a marker in the file header indicating an encoding change would be useful. But I fear that the information will not be reliable, because we only detect the encoding based on the added/deleted sections of the file. To detect the encoding reliably I think the complete old and new version of the file would need to be processed. The diff encoding detection is a best effort in order to attempt to render the diff correctly.

I would expect the commit message to mention the encoding change, providing the necessary context for the user to be able to interpret an apparently empty diff.

**Comment by 6543** [comment]
Didnt look to close but seeing 3 created buffers sound bit memory consuming in the first place

**Comment by jpraet** [comment]
> 
> 
> Didnt look to close but seeing 3 created buffers sound bit memory consuming in the first place

The 3 buffers are being reused for each file, and their combined content should be the same as the previous single buffer (just split out across 3 separate buffers by line type). So, at first sight, memory consumption looks similar to the previous version to me. But I'm new to golang, so I don't know for sure.

File: services/gitdiff/gitdiff.go [hunk 1]
```diff
@@ -886,32 +887,43 @@ parsingLoop:
 	// FIXME: There are numerous issues with this:
```
Comment (6543) [inline services/gitdiff/gitdiff.go]:
```suggestion
	// TODO: There are numerous issues with this:
```

**Review by 6543 — APPROVED** [review]

**Pull request merged**